Filters

mcosmin222 opened the pull request:
I created a framework for inserting filter graphs into Uncompressed sample providers and implemented an example audio filter class (which works out of the box).

You can pass filters using the configurations described in the link bellow by creating AvEffectDefinition objects

For example, you can insert an aecho effect like so

new AvEffectDefinition("aecho", "0.8:0.88:60:0.4"); 
Just remember to remove the "=" and what it is before it, so "aecho=0.8:0.88:60:0.4" becomes "0.8:0.88:60:0.4", otherwise you will get ugly errors.

Effects can be enabled or disabled during playback at will.
I have tested this branch a little, it has no obvious memory leaks or hard hitting errors.

I haven't done anything for videos yet (should be roughly similar to the existing AudioEffect class, but with different configurations for buffer source and resampler filter), and I am not sure if it is even possible for compressed samples.

https://ffmpeg.org/ffmpeg-filters.html#Audio-Filters

PS: I am not sure why the entire history of the repository is visible here, only the last 2 commits matter.

PS2: I know filters have several ways of initialization, but i find the string based approach the best, and it seems to work every time, and is easiest and most versatile to implement. The context options API doesn't seem to work every time. I only used it for the abuffer source because that filter has many options that would be harder to express as strings.

lukasf commented:
Nice work! I will need some time to play with this. But it sure looks promising.

mcosmin222 commented:
There is a little correction. The filter definition string should be without the "=". So the example above is

AvEffectDefinition("aecho", "0.8:0.88:60:0.4")

This gets me every time.

mcosmin222 commented:
Anyway, there are 2 things I am not 100% sure about so far:

1) Weather we need a mutex to protect the destructor of AudioEffect and the Add/Get frame methods

2) weather we need that flag with "m_PresetFilterPending" in UncompressedFrameProvider.

There are hypothetical chances of thread races there.

lukasf commented:
I have pushed a commit to your branch, I hope this is ok. I deleted the mutex you had and instead lock the mutex in FFmpegInterop whenever filters are changed. That way we make sure that no filters are changed while a sample is produced or a seek is ongoing. This should be slightly more effective. I am not sure what the m_PresetFilterPending was supposed to do, I would have deleted it as well.

I have deleted all empty .cpp files. They are not needed as long as the .h file is referenced somewhere.

These are some things that came to my mind when looking over the code:

1. Why do you declare some classes as "normal" classes and others as ref classes? I would propose to use ref classes throughout the library. They are much easier to use, no need to check result of new call (it throws when out of memory), no need for destructor unless you reference non-ref data, no manual ref counting or delete needed, no need to initialize fields (unless you want to have non-default value). The performance-critical stuff is done in ffmpeg, in the lib we can safely use ref classes without performance penalty.

2. What do you think about using specific AudioEffectDefinition and VideoEffectDefinition? This would allow for some optimizations, like specifying the output format of the effects. When we add the effect at at startup (maybe we can put them in config?), we can use this format for the descriptor and avoid resampler at the end of filter chain. Also if the required input format of effects is known, we can try to configure audio decoders to decode directly to that format (some codecs can output in different formats). This would avoid automatic resampling at start of filter chain. Maybe a structure like one AudioEffectsDefinition instance which contains a list of AvFilterDefinitions, plus the general parameters?

3. In the audio filter part, you use channel count and channel layout directly from format codex. AAC-HEv2 specifies 1 channel in context but actual output is 2 channels. It would be better to store the channel number and layout in the audio sample provider and use it to initialize audio filters as well.

mcosmin222 commented:
1) I just found these easier to use than ref classes. Sure, we can convert them to ref classes, there is no particular reason to use pure c++ classes above ref classes. 

2) I am fine with splitting them between audio and video if there are optimizations. I haven't found how you can set the specific input and output of effect without using some sort of resampler at the end of the chain. the audio abuffer source needs to know about the input, which is the same input you get from the avcodec context. The output is different, hence why the resampler at the end. I did not find how to explicity set the output format of random audio effects, I searched for a bit and haven't found anything useful in the docs. To what exactly are you referring?

3) Sure, I am all for it, I don't have such a test file. Can you do the honours please?

lukasf commented:
Regarding 2) I did not mean to change the output of effect filters. I think that cannot be changed. But if we play a file which is S16, and we know from start we will be using effects which output float, then we could configure the MSS as Float output. Then we do not need resampler at end of effect chain. Of course it only works when we know the effects during file open. If filters are changed at runtime and filter output changes, we will need resampling somewhere.

Also, one other idea I had was to dynamically create resampler in UncompressedAudioSampleProvider. So instead of creating it at file open, we could check every frame and create or re-create the resampler whenever we get a format change. If we had this, we could omit resampler filter, because we do resampling later, if required. This would also help with streams or TV rips, where sometimes the audio format of a stream changes during playback. I do not have such files, but I know it can happen.

mcosmin222 commented:
Actually, I was thinking about the dynamic resampler as well. I was just about to start on this after we got the filter. It is probably the better approach.

mcosmin222 commented:
Ahh I see now what you mean with the hevc2 format. I will see how to integrate this and dynamic resampling.

mcosmin222 commented:
Anyway, I added an experimental dynamic resampler and removed the resampler from the filter graph. It *works*. It is fairly trivial to test it with ordinary video and audio files.

The one thing this dynamic resampling doesn't do is check the channel layouts. This seems a little more complicated, because frames are apparently not aware of their profile, so to translate your channel check would mean mix check between AVcodecContext profile and AVFrame channel (layout). I am not sure this is valid, but I am pretty sure it is possible to have variance in the AVFrame channel layout - someone can add a resampler filter in the graph to resample it to a different channel layout. Can you have a look at it?

lukasf commented:
Cool, this looks good. I think we do not need to check the profile, becuase the AVFrame should already have the right channel number set. The decoder should detect the side format and add the second channel. At least that is how I think it works. But I will do some testing to verify this, maybe this evening or tomorrow.

lukasf commented:
I have pushed a new commit. I tried to simplify the format check, I think it should be more intuitive now, to understand what is happening. When testing with AAC-HE V2, I found that the current channel handling was wrong. This is also fixed now and works for all test files I have.

mcosmin222 commented:
It also works for all my files. I will give another nightly run (cycle for 8 hours between files of different formats) and run a profiler again to see if memory leaks and we are good to go I think.

lukasf commented:
Before pushing this, I'd like to change effect classes to ref classes. Until now, all classes in the lib are ref classes. I think it would be good to keep it like that, to have one uniform code style.

lukasf commented:
Good idea with a duration test, by the way.

mcosmin222 commented:
I will change them to ref classes.

lukasf commented:
Thank you

mcosmin222 commented:
I'd rather keep the delete calls than relaying on ref counting to delete the object. If we latter reference this object somewhere else, we will run into issues.

lukasf commented:
I'd say, if you reference the object somewhere else and then delete here, that is when you will run into problems. Ref classes are used throughout the lib and never explicitly deleted. This has not caused any issues.

mcosmin222 commented:
Well, I fired up the test - but with my version of the code. Fingers crossed ^^.

I will take your version and run it through the profile to see if we have any dangling objects. But there should be none, since it is logically equivalent to mine and mine has none.

mcosmin222 commented:
Regarding the delete, if someone references the code somewhere and then uses it after delete he will get a crash thats pretty easy to understand.
If he makes a reference somewhere and doesn't have the delete, he will have a hard to track memory leak.

The other objects are tied to the lifecycle of the ffmpeg interop object, whereas at least the effects themselves, objects here are created dynamically at playback time, and their lifetime is not always the same. Hence why I preferred to explicitly delete them, even when using ref classes: to tell people these objects must die here and now.

lukasf commented:
The classes are library internal. No one can create a reference to them from outside. If someone works on the library and creates a memory leak, well, then it is his problem. He can create memory leaks everywhere. Not using ref classes does not prevent one from making stupid coding mistakes. I'd say it is even the opposite: Using ref classes will reduce chances of having memory leaks, that is what ref classes is all about. The whole WinRT is based on ref classes for a good reason.

mcosmin222 commented:
There is a leak. Ill fix it soon.

mcosmin222 commented:
OK, it was a virtual destructor thingy again.
The only objects which don't get cleared right away are AvEffectDefinitions which remain referenced from managed code.

I wonder if we should delete them manually in unmanaged code or let the GC do its fun.

lukasf commented:
Ah this virtual destructor thing is a constant annoyance.

The effect definitions do not contain much data and they will be cleared up from GC at some point. Also the user might want to re-use the same definition for the next file as well, so in that case it would even be dangerous to delete them.

mcosmin222 commented:
We are good to go, no unmanaged resource leaks.

lukasf commented:
Alright, thanks for the great addition!
